lib/backends/sdc/containers.js

```diff
@@ -559,60 +525,302 @@ function addRulesToFWAPI(opts, rules, payload, callback) {
+                    fabric: true,
+                    owner_uuid: opts.account.uuid
+                };
+
+                log.debug({ listParams: listParams },
+                    format('Networks: searching for network %s',
+                        listParams.uuid));
+
+
+                napi.listNetworks(listParams,
+                    { headers: {'x-request-id': opts.req_id }},
+
+                    // we may have queried on a string shorter than
+                    // what the user supplied, and so the results may
+                    // include false positives - we need to drop them.
+                    function _postFilter(err, networks) {
+                        if (err) {
+                            return callback(err);
+                        }
+
+                        var _networks = networks.filter(function (network) {
+                            var netId;
+
+                            netId = utils.networkUuidToDockerId(network.uuid);
+
+                            return netId.substr(0, query.length) === query;
+                        });
+
+                        return cb(null, _networks);
+                    });
             }
+        ]
+    }, function _listedNetworks(err, results) {
+        if (err) {
+            return callback(new errors.napiErrorWrap(err,
+                format('error listing on query %s', query)));
+        }
+
+        // concat all results.
+        var networks = results.successes.reduce(function (acc, e) {
+            return acc.concat(e);
         });
 
-        if (!external_net) {
+        if (networks.length < 1) {
             callback(new errors.DockerError(
-                'unable to find "'+netName+'" network uuid'));
+                format('unable to find network matching %s', query)));
             return;
         }
 
-        payload.networks = [ {uuid: external_net, primary: true} ];
+        log.debug({ networks: networks },
+            'Networks: found candidate networks, sorting');
+
+        // pick the 'best' match, following Docker's heuristic that:
+        // 1. exact id is always best
+        // 2. exact name match
+        // 3. partial id match
+        // >1 partial id is an ambiguous result, but we do not know
+        // the order in which we're going to evaluate networks.
+        var network = networks.reduce(function (acc, nw, i, c) {
+
+            // what type have we already found?
+            var nameMatch = acc.name === query;
+            var idMatch = utils.networkUuidToDockerId(acc.uuid) === query;
+
+            // what type are we looking at?
+            var exactId = utils.networkUuidToDockerId(nw.uuid) === query;
+            var exactName = nw.name === query;
+
+            if (exactId) {
+                // always wins.
+                acc = nw;
+            } else if (exactName && !idMatch) {
+                // wins if we don't have an id match.
+                acc = nw;
+            } else if (acc && !idMatch && !nameMatch) {
```

Comment:
My apologies, `acc` is undefined unless a second parameter is passed to `reduce`.